### PR TITLE
v3.2 Changelog

### PR DESCRIPTION
**Pull request by esethna**
WIP by PMs

### CODE REVIEW

**Comment by lfbrock** [comment]
Here's a draft of my items (doesn't include bug fixes), not sure how we want to organize the categories: 

LDAP:
- Added a button to manually trigger LDAP synchronization
- Updating Synchronization Interval no longer requires a server restart to take effect
- Improved logging for LDAP synchronization
- Added validation to the LDAP settings in the System Console so an error is triggered if required fields are missing

Customization:
- Added the ability to create Custom Emoji (must be enabled from System Console > Customization > Custom Emoji)

Enterprise:
- Added the ability to restrict  the creation of Custom Emoji to administrators only
- Added the ability to set different themes per team

Web UI: 
- Added an online indicator to the header of Direct Message channels
- Added database type to the about dialog

**Comment by enahum** [comment]
I think you're missing SAML for Authentication in EE

**Comment by lfbrock** [comment]
Jason should be updating a list of his items, I think he'll include SAML

**Comment by jasonblais** [comment]
Yeah, draft about done, will post tonight

**Comment by jasonblais** [comment]
My draft:

**Highlights**

Languages:
 - Added German translation for user interface

SAML Single-Sign-On (Enterprise):
 - Users can sign-in to Mattermost with their SAML credentials
 - Mattermost accounts can display user information from SAML, such as first and last name, email and username
 - New Mattermost user accounts are automatically created the first time a user signs in with their SAML credentials on the Mattermost server.
 - Mattermost officially supports Okta and Microsoft ADFS as the identity providers (IDPs), but you may also try configuring SAML for a custom IDP.

**Improvements**

User Interface:
 - Removed unnecessary resizing when opening and closing the right hand sidebar
 - Removed jumping of the center channel when new messages are posted
 - Updated the channel info dialog to be more user friendly

Localization:
 - Added German translation for user interface
 - fixed v3.1 issue..?

Enterprise:
 - SAML authentication support with Okta and Microsoft ADFS
 - [New command line tools](http://docs.mattermost.com/administration/command-line-tools.html) such as inviting and removing users from channels, and restoring previously deleted channels.

**Bug Fixes:** 
 - Email body now contains the `siteURL` when inviting a user by email via CLI (command line interface)
 - YouTube videos now stop playing when collapsed.
 - Fixed error when adding an incoming webhook to a public channel the user is currently not in

**Compatibility Changes:**

Changes to Enterprise Edition:
 - Under `SamlSettings` in `config.json`:
   - Added `"Enable": false` to allow login using SAML. See [documentation](http://docs.mattermost.com/deployment/sso-saml.html) to learn more about configuring SAML for Mattermost.
   - Added `"Verify": false` to control whether Mattermost verifies the signature sent from the SAML Response matches the Service Provider Login URL.
   - Added 
   - Added 
   - Added 
   - Added 
   - Added 
   - Added 
   - Added 
   - Added

**Comment by jasonblais** [comment]
My draft:

**Highlights**

Languages:
 - Added German translation for user interface

SAML Single-Sign-On (Enterprise):
 - Users can sign-in to Mattermost with their SAML credentials
 - Mattermost accounts can display user information from SAML, such as first and last name, email and username
 - New Mattermost user accounts are automatically created the first time a user signs in with their SAML credentials on the Mattermost server.
 - Mattermost officially supports Okta and Microsoft ADFS as the identity providers (IDPs), but you may also try configuring SAML for a custom IDP.

**Improvements**

User Interface:
 - Removed unnecessary resizing when opening and closing the right hand sidebar
 - Removed jumping of the center channel when new messages are posted
 - Updated the channel info dialog to be more user friendly

Localization:
 - Added German translation for user interface

Enterprise:
 - SAML authentication support with Okta and Microsoft ADFS
 - [New command line tools](http://docs.mattermost.com/administration/command-line-tools.html) added, such as inviting and removing users from channels, and restoring previously deleted channels.

**Bug Fixes:** 
 - Fixed an issue where new languages are not added by default. Any server which upgraded to Mattermost v3.1 should set **System Console > Localization > Available Languages** blank to have new languages added by default.
 - Previously, a few shortcuts that used `CTRL` were overwriting existing messaging shortcuts in Mac. This has been changed so they only work with `CMD`. See [documentation](http://docs.mattermost.com/help/messaging/keyboard-shortcuts.html) for more details.
 - Email body now contains the `siteURL` when inviting a user by email via CLI (command line interface)
 - YouTube videos now stop playing when collapsed.
 - Fixed error when adding an incoming webhook to a public channel the user is currently not in

**Compatibility Changes:**

Changes to Team Edition:
 - Under `EmailSettings` in `config.json`:
   - Added `"FeedbackOrganization": ""` to specify organization name and address, which will be displayed on email notifications from Mattermost.

 - Under `LocalizationSettings` in `config.json`:
   - Changed: `"AvailableLocales": ""` to allow new languages be added by default.

 - Under `LogSettings` in `config.json`:
   - Added `"EnableWebhookDebugging": true`. When set to `true`, contents of incoming webhooks are printed to log files for debugging.

Changes to Enterprise Edition:
 - Under `EmailSettings` in `config.json`:
   - Added `"FeedbackOrganization": ""` to specify organization name and address, which will be displayed on email notifications from Mattermost.

 - Under `LocalizationSettings` in `config.json`:
   - Changed: `"AvailableLocales": ""` to allow new languages be added by default.

 - Under `LogSettings` in `config.json`:
   - Added `"EnableWebhookDebugging": true`. When set to `true`, contents of incoming webhooks are printed to log files for debugging.

 - Under `SamlSettings` in `config.json`:
   - Added `"Enable": false` to allow login using SAML. See [documentation](http://docs.mattermost.com/deployment/sso-saml.html) to learn more about configuring SAML for Mattermost.
   - Added `"Verify": false` to control whether Mattermost verifies the signature sent from the SAML Response matches the Service Provider Login URL.
   - Added `"Encrypt": false`to control whether Mattermost will decrypt SAML Assertions encrypted with your Service Provider Public Certificate.
   - Added `"IdpUrl": ""` to set the SAML SSO URL where Mattermost sends a SAML request to start login sequence.
   - Added `"IdpDescriptorUrl": ""` to set the Identity Provider Issuer URL for the Identity Provider you use for SAML requests.
   - Added `"AssertionConsumerServiceURL": ""` to set the Service Provider Login URL.
   - Added `"IdpCertificateFile": ""` to set the public authentication certificate issued by your Identity Provider.
   - Added `"PublicCertificateFile": ""` to set certificate used to generate the signature on a SAML request to the Identity Provider for a service provider initiated SAML login, when Mattermost is the Service Provider.
   - Added `"PrivateKeyFile": ""` to set the private key used to decrypt SAML Assertions from the Identity Provider.
   - Added `"FirstNameAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the first name of users in Mattermost.
   - Added `"LastNameAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the last name of users in Mattermost.
   - Added `"EmailAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the email of users in Mattermost.
   - Added `"UsernameAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the username of users in Mattermost.
   - Added `"NicknameAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the nickname of users in Mattermost.
   - Added `"LocaleAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the language of users in Mattermost.
   - Added `"LoginButtonText": ""` set the text that appears in the login button on the login page.

**Comment by jasonblais** [comment]
There are a few config updates from the emoji PRs (https://github.com/mattermost/platform/pull/3381, https://github.com/mattermost/platform/pull/3311), but I wasn't sure if @lfbrock you were working on adding the updates.

If not, I can just add them to my section.

Same with https://github.com/mattermost/platform/pull/3442 @esethna

**Comment by lfbrock** [comment]
Ah I'm not sure, I can add them but it might be easier to have the release manager manage the config.json/compatibility sections and bug fixes, and the PMs write the improvements?

0/5, was just thinking since those are more comparing lists it might be easiest to have one person do them so things don't get missed.

**Comment by jasonblais** [comment]
I'm 0/5 as well, we can have the release manager do config.json/compatibility sections and bug fixes for this release (as before).

I've just been adding changelog items to my list whenever I've added a config change or tested a bug. 

That might be easier in the long term so we have each PM tracking their own issues/tickets, then release manager confirming nothing was missed.

File: source/administration/changelog.md [hunk 1]
```diff
@@ -2,6 +2,224 @@
...
+- Testing the email connection in the System Console no longer throws an error.
+- Multiline list items are now displayed correctly on new lines.
+- Error message is updated when switching from email to GitLab SSO authentication that is already used by another account.
+- Timestamps no longer require a page refresh when switching between 12h and 24h display formats.
+- `#in`, `#from`, or `#channel` now highlight in search results.
+- No longer require a page refresh before enabling compliance reporting in the System Console.
+- `@all` no longer sends mentions if unselected in Account Settings.
+- Users are no longer redirected to the switch teams page after changing authentication method from GitLab SSO to email.
+- Invalid MFA token error message now clears correcly from the UI.
+- Errors now correcly clear from the UI when changing passwords.
```
Comment (lfbrock) [inline source/administration/changelog.md]: correcly > correctly

File: source/administration/changelog.md [hunk 2]
```diff
@@ -2,6 +2,224 @@
...
+- System Console users list no longer throws an error when trying to demote a member from a System Admin.
+- iOS radio buttons no longer stay selected when switching between options.
+- Email addresses now display for System Admins even if hidden in the System Console.
+- Code themes now save when updated via Account settings.
+- file name is now displayed inctead of the full path to the file in code snippet previews.
```
Comment (lfbrock) [inline source/administration/changelog.md]: file > File

File: source/administration/changelog.md [hunk 3]
```diff
@@ -2,6 +2,224 @@
...
+### Highlights
+
+#### Custom Emoji
+
+- Create Custom Emoji from the **Main Menu** > **Custom Emoji** when enabled from the **System Console** > **Customization** > **Custom Emoji**.
+- Restrict the permissions required to create Custom Emoji (Enterprise).
+
+#### SAML Single-Sign-On (Enterprise):
+
+- New Mattermost user accounts are automatically created the first time a user signs in with their SAML credentials. Mattermost pulls user information from SAML, including first and last name, email and username.
```
Comment (jasonblais) [inline source/administration/changelog.md]: Propose adding `Users can sign in to Mattermost with their SAML credentials`, e.g.:

`Users can sign in to Mattermost with their SAML credentials and new Mattermost user accounts are automatically created on first login. Mattermost pulls user information from SAML, including first and last name, email and username.`

File: source/administration/changelog.md [hunk 4]
```diff
@@ -2,6 +2,224 @@
...
+#### Languages
+
+- Added German translation for the user interface if enabled by the System Admin from **System Console > Localization > Available Languages**.
+
+
+### Improvements
+
+**On-Boarding**
+
+- After account creation, users are automatically directed to the team where they were invited instead of the Select Teams page.
```
Comment (jasonblais) [inline source/administration/changelog.md]: `Team Selection` page

File: source/administration/changelog.md [hunk 5]
```diff
@@ -2,6 +2,224 @@
...
+- Errors now correcly clear from the UI when changing passwords.
+- System Console users list no longer throws an error when trying to demote a member from a System Admin.
+- iOS radio buttons no longer stay selected when switching between options.
+- Email addresses now display for System Admins even if hidden in the System Console.
+- Code themes now save when updated via Account settings.
+- file name is now displayed inctead of the full path to the file in code snippet previews.
+- Config settings are refreshed immediately when **Reload Configuration from Disk** is clicked.
+- Preview feature checkboxes now reset after changes are canceled.
+- Updated the markdown parser to fix poor handling of certain links.
+- Right-hand sidebar no longer appears black in some cases on IE11 and Edge.
```
Comment (lfbrock) [inline source/administration/changelog.md]: Was this in v3.1? Can we do a double check of the bugs list and remove anything that wasn't actually a bug in v3.1, and was only introduced afterwards?

File: source/administration/changelog.md [hunk 6]
```diff
@@ -2,6 +2,224 @@
...
+- “More” option under Direct Message list no longer shows count of team members not in your direct message list.
+- On Firefox, search results for hashtags are not properly highlighted.
+- Clicking on a desktop notification from another team doesn’t open the team.
+- Webhook attachments don't show up in search results.
+- On Firefox, System Console sidebar completely disappears when an LDAP setting is saved
+- On Firefox, `CTRL/CMD + U` keyboard shortcut doesn't work
+- Copying and pasting an image from a browser doesn't work
+- `/join` sometimes throws an error
+- When upgrading to 3.X, syntax highlighting using Solarized code theme is lost
+- In Compact view, clicking on a file in the first post in the right hand sidebar attempts to download the file
```
Comment (jasonblais) [inline source/administration/changelog.md]: I think this is fixed

File: source/administration/changelog.md [hunk 7]
```diff
@@ -2,6 +2,224 @@
...
+
+### Known Issues
+
+- “More” option under Direct Message list no longer shows count of team members not in your direct message list.
+- On Firefox, search results for hashtags are not properly highlighted.
+- Clicking on a desktop notification from another team doesn’t open the team.
+- Webhook attachments don't show up in search results.
+- On Firefox, System Console sidebar completely disappears when an LDAP setting is saved
+- On Firefox, `CTRL/CMD + U` keyboard shortcut doesn't work
+- Copying and pasting an image from a browser doesn't work
```
Comment (jasonblais) [inline source/administration/changelog.md]: Fixed (unless we want to add a note it doesn't work for Firefox and IE11)

File: source/administration/changelog.md [hunk 8]
```diff
@@ -2,6 +2,224 @@
...
+- `/join` sometimes throws an error
+- When upgrading to 3.X, syntax highlighting using Solarized code theme is lost
+- In Compact view, clicking on a file in the first post in the right hand sidebar attempts to download the file
+- Unable to leave a private channel in mobile view
```
Comment (jasonblais) [inline source/administration/changelog.md]: fixed

File: source/administration/changelog.md [hunk 9]
```diff
@@ -2,6 +2,224 @@
...
+- Invalid config setting causes server to panic on start
```
Comment (jasonblais) [inline source/administration/changelog.md]: I think it's fixed with https://github.com/mattermost/platform/pull/3361

File: source/administration/changelog.md [hunk 10]
```diff
@@ -2,6 +2,215 @@
...
+- Added validation to the LDAP settings in the System Console so an error is triggered if required fields are missing.
+
+**Enterprise**
+
+- [New command line tools](http://docs.mattermost.com/administration/command-line-tools.html) added, such as inviting and removing users from channels, and restoring previously deleted channels.
+- Added the ability to set different themes per team.
+
+### Bug Fixes
+
+- Smileys ending in a letter are now prioritized first in the autocomplete list.
```
Comment (jasonblais) [inline source/administration/changelog.md]: sounds like an improvement

File: source/administration/changelog.md [hunk 11]
```diff
@@ -2,6 +2,224 @@
...
+
+- “More” option under Direct Message list no longer shows count of team members not in your direct message list.
+- On Firefox, search results for hashtags are not properly highlighted.
+- Clicking on a desktop notification from another team doesn’t open the team.
+- Webhook attachments don't show up in search results.
+- On Firefox, System Console sidebar completely disappears when an LDAP setting is saved
+- On Firefox, `CTRL/CMD + U` keyboard shortcut doesn't work
+- Copying and pasting an image from a browser doesn't work
+- `/join` sometimes throws an error
+- When upgrading to 3.X, syntax highlighting using Solarized code theme is lost
```
Comment (lfbrock) [inline source/administration/changelog.md]: I think this was fixed, maybe check with DL? https://mattermost.atlassian.net/browse/PLT-3150

File: source/administration/changelog.md (hunk 7 above)
Comment (lfbrock) [inline source/administration/changelog.md]: This was fixed, (for Chrome + Edge, and it is not a supported feature for other browsers so not really a bug).

File: source/administration/changelog.md [hunk 12]
```diff
@@ -2,6 +2,215 @@
...
+- [New command line tools](http://docs.mattermost.com/administration/command-line-tools.html) added, such as inviting and removing users from channels, and restoring previously deleted channels.
+- Added the ability to set different themes per team.
+
+### Bug Fixes
+
+- Smileys ending in a letter are now prioritized first in the autocomplete list.
+- Clarified the GitLab SSO error message if another Mattermost account is already associated with the GitLab account.
+- Privacy settings in the system console now refresh correctly when hiding email addresses or full names.
+- Fixed the cross contamination of new channels created on different teams in the same browser.
+- Create channel modal now saves upon pressing `ENTER` or `CTRL+ENTER` depending on the user setting.
```
Comment (jasonblais) [inline source/administration/changelog.md]: works in v3.1

File: source/administration/changelog.md (hunk 5 above)
Comment (jasonblais) [inline source/administration/changelog.md]: Not a problem in v3.1

File: source/administration/changelog.md (hunk 6 above)
Comment (lfbrock) [inline source/administration/changelog.md]: This seems fixed

File: source/administration/changelog.md (hunk 8 above)
Comment (lfbrock) [inline source/administration/changelog.md]: This also seems fixed

File: source/administration/changelog.md [hunk 13]
```diff
@@ -2,6 +2,209 @@
...
+**Enterprise**
+
+- [New command line tools](http://docs.mattermost.com/administration/command-line-tools.html) added, such as inviting and removing users from channels, and restoring previously deleted channels.
+- Added the ability to set different themes per team.
+
+### Bug Fixes
+
+- Privacy settings in the system console now refresh correctly when hiding email addresses or full names.
+- Fixed the cross contamination of new channels created on different teams in the same browser.
+- Create channel modal now saves upon pressing `ENTER` or `CTRL+ENTER` depending on the user setting.
```
Comment (jasonblais) [inline source/administration/changelog.md]: not a bug in v3.1

File: source/administration/changelog.md [hunk 14]
```diff
@@ -2,6 +2,224 @@
...
+- Added `DeleteAt` column.
+
+**Emoji Table**    
+- Added `Emoji` table.
+
+### Known Issues
+
+- “More” option under Direct Message list no longer shows count of team members not in your direct message list.
+- On Firefox, search results for hashtags are not properly highlighted.
+- Clicking on a desktop notification from another team doesn’t open the team.
```
Comment (lfbrock) [inline source/administration/changelog.md]: Seems fixed

File: source/administration/changelog.md [hunk 15]
```diff
@@ -2,6 +2,224 @@
...
+**TeamMembers Table**    
+- Added `DeleteAt` column.
+
+**Emoji Table**    
+- Added `Emoji` table.
+
+### Known Issues
+
+- “More” option under Direct Message list no longer shows count of team members not in your direct message list.
+- On Firefox, search results for hashtags are not properly highlighted.
```
Comment (lfbrock) [inline source/administration/changelog.md]: Seems fixed

File: source/administration/changelog.md (hunk 9 above)
Comment (lfbrock) [inline source/administration/changelog.md]: I think this is fixed, maybe check with EN https://mattermost.atlassian.net/browse/PLT-3364

File: source/administration/changelog.md [hunk 16]
```diff
@@ -2,6 +2,209 @@
...
+- Mattermost officially supports Okta and Microsoft ADFS as the identity providers (IDPs), but you may also try configuring SAML for a custom IDP.
+
+#### Policy (Enterprise)
+
+- Restrict the permission levels required to send team invitiations in **System Console** > **Policy**.
+- Restrict the permission levels required to manage public and private channels, including creating, deleting, renaming, and setting the channel header or purpose. 
+
+#### Languages
+
+- Added German translation for the user interface if enabled by the System Admin from **System Console > Localization > Available Languages**.
```
Comment (esethna) [inline source/administration/changelog.md]: move first

File: source/administration/changelog.md [hunk 17]
```diff
@@ -2,6 +2,209 @@
...
+**Notifications**
+
+- Address displayed in the email notification footer is now configurable in the System Console.
+- Direct Message desktop notifications now display with a "Direct Message" title.
+
+**Web UI**
+
+- Reply button and [...] menu now appear in a hovering UI element to increase the available margin width in the center channel.
+- Right-hand sidebar can now be expanded when viewing threads or search results.
+- Smileys ending in a letter are now prioritized first in the autocomplete list.
```
Comment (esethna) [inline source/administration/changelog.md]: check the bug

File: source/administration/changelog.md [hunk 18]
```diff
@@ -2,6 +2,209 @@
...
+
+- Restrict the permission levels required to send team invitiations in **System Console** > **Policy**.
+- Restrict the permission levels required to manage public and private channels, including creating, deleting, renaming, and setting the channel header or purpose. 
+
+#### Languages
+
+- Added German translation for the user interface if enabled by the System Admin from **System Console > Localization > Available Languages**.
+
+
+### Improvements
```
Comment (esethna) [inline source/administration/changelog.md]: Remove users from team, "adding" to a channel

File: source/administration/changelog.md [hunk 19]
```diff
@@ -2,6 +2,209 @@
...
+
+- Create Custom Emoji from the **Main Menu** > **Custom Emoji** when enabled from the **System Console** > **Customization** > **Custom Emoji**.
+- Restrict the permissions required to create Custom Emoji (Enterprise).
+
+#### SAML Single-Sign-On (Enterprise):
+
+- Users can sign in to Mattermost with their SAML credentials and new Mattermost user accounts are automatically created on first login. Mattermost pulls user information from SAML, including first and last name, email and username.
+- Mattermost officially supports Okta and Microsoft ADFS as the identity providers (IDPs), but you may also try configuring SAML for a custom IDP.
+
+#### Policy (Enterprise)
```
Comment (esethna) [inline source/administration/changelog.md]: Policy above

File: source/administration/changelog.md [hunk 20]
```diff
@@ -2,6 +2,209 @@
...
+- Under `ServiceSettings` in `config.json`:
+   - Added `"EnableCustomEmoji": false`. When set to `true`, enables Custom Emoji option in the Main Menu where users can create customized emoji.
+
+- Under `LocalizationSettings` in `config.json`:
+   - Changed: `"AvailableLocales": ""` to allow new languages be added by default.
+
+- Under `LogSettings` in `config.json`:
+   - Added `"EnableWebhookDebugging": true`. When set to `true`, contents of incoming webhooks are printed to log files for debugging.
+
+**Changes to Enterprise Edition:**
```
Comment (esethna) [inline source/administration/changelog.md]: gzip, download time, maybe add to highlights, webpack splitting ("Performance")

File: source/administration/changelog.md [hunk 21]
```diff
@@ -2,6 +2,209 @@
...
+   - Added `"LocaleAttribute": ""` to set the attribute in the SAML Assertion that will be used to populate the language of users in Mattermost.
+   - Added `"LoginButtonText": ""` set the text that appears in the login button on the login page.
+
+#### Database Changes from v3.1 to v3.2
+
+**TeamMembers Table**    
+- Added `DeleteAt` column.
+
+**Emoji Table**    
+- Added `Emoji` table.
```
Comment (esethna) [inline source/administration/changelog.md]: Remove per team themes from changelog

File: source/administration/changelog.md [hunk 22]
```diff
@@ -2,6 +2,215 @@
...
+- Privacy settings in the system console now refresh correctly when hiding email addresses or full names.
+- Fixed the cross contamination of new channels created on different teams in the same browser.
+- Updated the GitLab SSO error message for clarity if another Mattermost account is already associated with the GitLab account.
+- Team creation via GitLab SSO no longer throws an error if email domains are restricted.
+- Channel header no longer disappears after renaming a channel
+- Testing the email connection in the System Console no longer throws an error.
+- Multiline list items are now displayed correctly on new lines.
+- Error message is updated when switching from email to GitLab SSO authentication that is already used by another account.
+- Timestamps no longer require a page refresh when switching between 12h and 24h display formats.
+- `#in`, `#from`, or `#channel` now highlight in search results.
```
Comment (lfbrock) [inline source/administration/changelog.md]: This should be in hashtags **starting** with #in, #from, #channel (#in is not a valid hashtag, and #from seems to bring up no results)

File: source/administration/changelog.md [hunk 23]
```diff
@@ -2,6 +2,215 @@
...
+- Updated labeling of System Console settings in the UI for consistency and accuracy.
+- (Enterprise) Help text was added to the **Reload Configuration from Disk** and **Recycle Database Connections** buttons to explain their use in the switching databases with no down time.
+- Allow System Admins to create teams even if team creation is disabled via the System Console.
+
+**Notifications**
+
+- Address displayed in the email notification footer is now configurable in the System Console.
+- Direct Message desktop notifications now display with a "Direct Message" title.
+
+**Web UI**
```
Comment (lfbrock) [inline source/administration/changelog.md]: Thought of another one - not sure if it fits in web ui?

@mention autocomplete now filters on nickname, full name, and username

**Comment by esethna** [comment]
Merging the changelog after making the tweaks suggested in Code Complete meeting and from PM reviews.

**Pull request merged**